### PR TITLE
[WIP] Basic implementation

### PR DESCRIPTION
**Pull request by Danappelxx**
cc @paulofaria

### CODE REVIEW

File: Sources/main.swift [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+//MARK: Resources
+let todoResource = makeTodoResource(store: store)
+
+//MARK: Main router
+let router = Router(middleware: cors) { route in
+    route.resources(path: "/", resources: todoResource)
+}
+
+//MARK: Server
```
Comment (paulofaria) [inline Sources/main.swift]: Don't think the `//MARK:`s are necessary here.. so little code

File: Sources/CORSMiddleware.swift [hunk 2]
```diff
@@ -0,0 +1,17 @@
+//
+//  CORSMiddleware.swift
+//  TODOBackend
+//
+//  Created by Dan Appel on 5/1/16.
+//
+//
+
```
Comment (paulofaria) [inline Sources/CORSMiddleware.swift:24]: Please change these to MIT headers retaining your copyright, of course.

File: Sources/Todo.swift [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+//
+//  Created by Dan Appel on 5/1/16.
+//
+//
+
+import Mapper
+import Resource
+
+struct Todo {
+    let id: Int?
```
Comment (paulofaria) [inline Sources/Todo.swift]: I wouldn't use `id` as optional. But this is personal choice haha.

File: Sources/TodoResource.swift [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+            return Response(content: todos.content)
+        }
+
+        // POST /
+        todo.create(content: Todo.self) { request, todo in
+            let id = todoStore.nextId()
+            let newTodo = Todo(
+                id: id,
+                url: "http://01aa3281.ngrok.io/\(id)",
+                title: todo.title, completed: todo.completed, order: todo.order)
```
Comment (paulofaria) [inline Sources/TodoResource.swift]: I would add a new line

``` swift
let newTodo = Todo(
  id: id,
  url: "http://01aa3281.ngrok.io/\(id)",
  title: todo.title,
  completed: todo.completed,
  order: todo.order
)
```

File: Sources/TodoResource.swift [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+
+        // DELETE /
+        todo.clear { request in
+            let deleted = todoStore.clear()
+            return Response(content: deleted.content)
+        }
+
+        // GET /:id
+        todo.show { request, id in
+            guard let id = Int(id), todo = todoStore.get(id: id) else {
```
Comment (paulofaria) [inline Sources/TodoResource.swift]: this `Int(id)` will soon be unnecessary 😈

File: Sources/TodoResource.swift (hunk 4 above)
Comment (Danappelxx) [inline Sources/TodoResource.swift]: It's probably better to use lenses here anyway.

Comment (paulofaria) [inline Sources/TodoResource.swift]: but this is an initializer.. what would it look like with lenses?

Comment (Danappelxx) [inline Sources/TodoResource.swift]:
``` swift
let newTodo = todo
    .with(id: id)
    .with(url: "http://01aa3281.ngrok.io/\(id)")
```

File: Sources/Todo.swift (hunk 3 above)
Comment (paulofaria) [inline Sources/Todo.swift]: Is there a reason it is `Optional`?

File: Sources/TodoResource.swift (hunk 4 above)
Comment (paulofaria) [inline Sources/TodoResource.swift]: Yeah this is the part I would do differently. You're treating the data you're receiving to construct a `Todo` as a `Todo` itself. But this is not a `Todo` what you're receiving is the information needed to construct a `Todo`. I would make `Todo` a `Struct` with no `Optional`s at all. I would create another `Struct`, something like `TodoInfo` (better name needed). And this `TodoInfo` would also not have any `Optional`.  I would also abstract the `id` generation inside `todoStore.insert(todo: newTodo)`. The `todoStore` would be responsible to receive `todoInfo` and return `todo` with the correct `id`.

**Comment by Danappelxx** [comment]
Updated to use 0.7, passes all tests right now.

Things to do
- [x] Clean up `Todo` (use a middleman struct as you suggested)
- [ ] Use PostgreSQL as the data store (currently using in-memory)
- [x] Move some of the StructuredData extensions in `TodoResource.swift` to StructuredData
- [x] Abstract the url (get it from environmental variables?)
- [x] Make file headers MIT

cc @paulofaria

**Comment by Danappelxx** [comment]
We can do PostgreSQL later. I'll merge this for now 🎉

**Pull request merged**